lib/engine/game/g_1824/game.rb

```diff
@@ -557,251 +579,297 @@ def sorted_corporations
+
+          # 1824 ADD BEGIN
+          float_corporation(corporation) if regional?(corporation) && corporation.floatable && floated != corporation.floated?
+          # 1824 ADD END
         end
 
-        def revenue_str(route)
-          # TODO: Implement Bukowina bonus if Cislethania map used
+        def place_home_token(corporation)
+          # When Staatsbahn is formed it uses existing Pre-Staatsbahn so no new token is placed
+          return if staatsbahn?(corporation)
 
           super
         end
 
-        def mine_revenue(routes)
-          routes.sum { |r| r.stops.sum { |stop| mine_hex?(stop.hex) ? stop.route_revenue(r.phase, r.train) : 0 } }
+        def associated_coal_railway(regional_railway)
+          coal_railway =
+            case regional_railway.name
+            when 'BK'
+              'EPP'
+            when 'MS'
+              'EOD'
+            when 'CL'
+              'MLB'
+            when 'SB'
+              'SPB'
+            end
+          corporation_by_id(coal_railway)
+        end
+
+        def revenue_for(route, stops)
+          super + bukowina_bonus_amount(route, stops)
+        end
+
+        def revenue_str(route)
+          str = super
+          str += ' + Bukowina' if bukowina_bonus_amount(route, route.stops).positive?
+          str
         end
 
         def float_str(entity)
-          return super if !entity.corporation || entity.floatable
+          return super unless entity.corporation
 
           case entity.id
           when 'BK', 'MS', 'CL', 'SB'
             needed = entity.percent_to_float
-            needed.positive? ? "#{entity.percent_to_float}% (including exchange) to float" : 'Exchange to float'
+            if needed.positive?
+              need_exchange = entity.floatable ? '' : ' + exchange'
+              "#{entity.percent_to_float}%#{need_exchange} to float"
+            else
+              'Exchange to float'
+            end
           when 'UG'
-            'U1 exchange floats'
+            'UG1 exchange floats'
           when 'KK'
-            'K1 exchange floats'
+            'KK1 exchange floats'
           when 'SD'
-            'S1 exchange floats'
+            'SD1 exchange floats'
           else
-            'Not floatable'
+            super
           end
         end
 
-        def all_corporations
-          @corporations.reject(&:removed)
-        end
-
-        def event_close_mountain_railways!
-          @log << '-- Any remaining Mountain Railways are either exchanged or discarded'
-          # If this list contains any companies it will trigger an interrupt exchange/pass step
-          @forced_mountain_railway_exchange = @companies.select { |c| mountain_railway?(c) && !c.closed? }
+        # 1824 uses standard sold out stock movement. Rule VIII.3, bullet 4: move up 1 if not at top
+        def sold_out_stock_movement(corporation)
+          Base.instance_method(:sold_out_stock_movement).bind_call(self, corporation)
```

Comment:
See my other `bind_call` comment; we should add a non-base method in G1837 to check if its special implementation should be used, and G1824 can override that to false so that it doesn't have to use `bind_call`.